### PR TITLE
Pre-zync synchronization fixes

### PR DESCRIPTION
**Pull request by andyross**
Pushing up some early fixes from the Zync work to make review easier for the final PR.  These are all genuine mixups, bugs or misfeatures in the existing code that don't have anything to do with zync other than that's how they were discovered.  In most cases the impact seems to have been benign, though the broken locking in the network and bluetooth code looks like the kind of thing security folks would worry about.

They aren't otherwise related to each other, but I have them all in one PR for my own convenience.  Hopefully they won't be controversial or involve much discussion, but if there's desire these can be split up as needed.

### CODE REVIEW

**Comment by andyross** [comment]
Oh, and if it's not clear: these patches mostly stand alone.  Don't try to review just looking at the single global diff, it won't make any sense.

**Review by dcpleung — DISMISSED** [review]

**Review by npitre — DISMISSED** [review]

**Review by peter-mitsis — DISMISSED** [review]

**Review by cfriedt — DISMISSED** [review]

**Comment by nashif** [comment]
@andyross  would be great if you can  present zync in the architecture WG and give an overview of the changes you are proposing. (talk to @carlescufi please)

**Comment by cfriedt** [comment]
@andyross - My obvious question - are there any bugfixes here? Obvious next question: if so, do they merit a backport? :-)

**Comment by andyross** [comment]
@nashif: definitely, it's just waiting on me to finish a PR that looks final-ish.  The long tail of stuff to fix was **really** long, we do weird things with our synchronization primitives.  Note that either the arch or dev meeting should be OK -- if I did it correctly, this doesn't actually change any APIs (though it does change the default semantics, you can tune back the original behavior via kconfig).

@cfriedt Almost all of them are putative bugs, though maybe low priority.  The ones I'd look at backporting first are the ones where it turned out synchronization was broken due to k_mutex errors being ignored.  Whether it's worth backporting CMSIS fixes... I dunno.  I guess if someone had actually used that to port a real app, we'd know about it already.  And if they haven't, they should be working against current code.

File: subsys/net/lib/dns/resolve.c
Comment (jukkar) [inline subsys/net/lib/dns/resolve.c:1429]: Hmm, the `dns_resolve_init()` is normally called just fine during system initialization by `dns_init_resolver()` in `subsys/net/ip/net_core.c`. But that DNS init might be skipped if DHCPv4 is not properly initialized, I suspect that something like this happens here.
The `dns_resolve_init()` in `dns_init_resolver()` looks like this
```
        ret = dns_resolve_init(dns_resolve_get_default(), dns_servers, NULL);
        if (ret < 0) {
                NET_WARN("Cannot initialize DNS resolver (%d)", ret);
        }
```
It looks like there would be issues here if we do it like this.

File: subsys/net/lib/dns/resolve.c
Comment (jukkar) [inline subsys/net/lib/dns/resolve.c:75]: Typo in the commit message `net/dns: Initialize default DNS context object`
s/dns_resovle_/dns_resolve/

**Review by jukkar — CHANGES_REQUESTED** [review]

**Comment by jukkar** [comment]
@andyross About the DNS fix, how did you manage to trigger it? After looking the DHCPv4 init code, it could only fail if UDP port registration fails, in which case the system network setup is probably wrong and we should not continue.
On the other hand If the `dns_init_resolver()` is not yet called and some entity manages to call a DNS function that acquires the lock, then we have a network init ordering problem that needs to be fixed elsewhere.

**Comment by Thalley** [comment]
@andyross For the ISO changes you write:

>  bluetooth: Don't use k_sem internals
>
> The "limit" field of struct k_sem was in use in two places in the BTLE host code, in one it was being used as a duplicate placeholder for the "iso_max_num" field received in read_buffer_size_v2_complete[1].  In another, it was just being tested for zero[2].
> 
> Those are pretty clear abuses of internal data, provide minimal value beyond a few bytes of memory in struct bt_dev_le, and in any case won't work with zync, where that field doesn't have the same name and may not even exist depending on app configuration.

Can you elaborate on what you mean by "internal data"? The `k_sem.limit` is publicly available in `kernel.h`, and the only indication that it is internal is the doxygen part
```
/**
 * @cond INTERNAL_HIDDEN
 */
```
Are you claiming that this is enough to make it "internal", and shouldn't be access by the BT stack? 

I suppose the same argument can be used for the `count` field, right? However for that field, we have `k_sem_count_get`. Couldn't we just implement a `k_sem_limit_get`, and still use the `k_sem` for hold the initial value, instead of duplicating data?

File: subsys/net/lib/dns/resolve.c
Comment (andyross) [reply subsys/net/lib/dns/resolve.c:1429]: What's the proper way to fix this, then?  You can see this pretty trivially in native_posix with tests/net/dhcpv4, just run it under gdb and set a breakpoint right here and on dns_resolve_init() to see that the latter never gets called.

**Comment by andyross** [comment]
> Are you claiming that this is enough to make it "internal", and shouldn't be access by the BT stack?

Pretty much.  The nature of C is that structs have to appear in public headers most of the time for efficiency reasons.  If a struct field doesn't appear in generated documentation, you shouldn't use it.

**Comment by Thalley** [comment]
> > Are you claiming that this is enough to make it "internal", and shouldn't be access by the BT stack?
> 
> Pretty much. The nature of C is that structs have to appear in public headers most of the time for efficiency reasons. If a struct field doesn't appear in generated documentation, you shouldn't use it.

That is a fair point. 

Would it be useful to implemt the `k_sem_limit_get` as I suggested? Clearly there's a use case for it (although only one at this point).

**Comment by andyross** [comment]
> Would it be useful to implement the k_sem_limit_get as I suggested? Clearly there's a use case for it (although only one at this point).

In principle, sure.  But in practice I don't know that I see value.  This semaphore limit is a weird feature of Zephyr that other implementations don't support.  It's confusing, because it looks like it should be a blocking counter (like a msgq, say) but what it really does is **saturate** (i.e. it drops the k_sem_give() on the floor and forgets about it, which is very unlikely to be what synchronization code wants!).  

The real reason I don't want this being used, basically, is that I want it to go away.  With zync there is a CONFIG_ZYNC_MAX_VAL tunable that is required for this feature to exist.  Because this requires semaphore internal state, it can't be implemented entirely in userspace.

And as mentioned in the commit message, I don't really see it being needed here.  The code isn't using it to try to detect semaphore state or do anything involving synchronization, it's just using it as a clever place to store the packet count it got at initialization time (and also, seemingly, to see if its container struct has been initialized!).  That's app data, not semaphore data, and it belongs in the app.

File: subsys/bluetooth/host/hci_core.h
Comment (Thalley) [inline subsys/bluetooth/host/hci_core.h:246]: The value stored here is the `rp->iso_max_num` which is a `uint8_t`. 

```suggestion
	uint8_t		iso_limit;
```

**Review by Thalley — CHANGES_REQUESTED** [review]

**Comment by andyross** [comment]
> > Are you claiming that this is enough to make it "internal", and shouldn't be access by the BT stack?
> 
> Pretty much. The nature of C is that structs have to appear in public headers most of the time for efficiency reasons. If a struct field doesn't appear in generated documentation, you shouldn't use it.

I should soften that a little.  I'm not a zealot about this.  In-tree code that can do something clever with other in-tree internals isn't a bad thing at all, and it tells us things about how the code "should" work vs. how it was intended.  It just needs to be aware that toes are being stepped on and that it can become a maintenance hassle as the internal code being used evolves.

File: subsys/net/lib/dns/resolve.c
Comment (jukkar) [reply subsys/net/lib/dns/resolve.c:1429]: @andyross I just submitted a fix proposal at https://github.com/zephyrproject-rtos/zephyr/pull/51267
You could just take the fix commit from there or wait until until it gets merged.

**Comment by andyross** [comment]
Fix typo, and remove the DNS patch in favor of @jukkar's in #51267

**Review by npitre — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Review by jukkar — APPROVED** [review]

**Review by Thalley — APPROVED** [review]

**Comment by cfriedt** [comment]
@rlubos If you have no concerns with this diff, please ack

**Comment by andyross** [comment]
Just noticed the failures in this rev.  This was passing earlier and the only changes are syntactic.  I can reproduce a build failure locally (mutiple definitions of "fff") but not the failure seen.  Though I'm not familiar with the unit_testing layer and may have a dependency missing or something.  Is this an upstream glitch?  @nashif @stephanosio ?

**Comment by cfriedt** [comment]
> Just noticed the failures in this rev. This was passing earlier and the only changes are syntactic. I can reproduce a build failure locally (mutiple definitions of "fff") but not the failure seen. Though I'm not familiar with the unit_testing layer and may have a dependency missing or something. Is this an upstream glitch? @nashif @stephanosio ?

We're seeing no failures in CI @andyross - can you specify which one is giving you problems locally?

**Comment by andyross** [comment]
Uh... I have no idea.  There was a failure in one of the bluetooth unit_testing tests!  I swear, it was right there.  But now it's not.  Never mind, I guess. :)

**Review by rlubos — APPROVED** [review]

**Pull request merged**
